### DISCUSSION TITLE
[基础教程2]如何使用AnyKernel 3打包内核

### DISCUSSION THREAD

**Discussion by Coconutat** [tiann/KernelSU#952] [Show and tell]
# 如何使用AnyKernel 3打包内核
正如标题，这依然是基础教程。经验丰富朋友依然可以无视。
  
***  
  
## 1.AnyKernel 3是什么？
[AnyKernel3](https://github.com/osm0sis/AnyKernel3)是由XDA论坛的开发人员[osm0sis](https://github.com/osm0sis)开发的刷写安卓设备内核的刷机模板。事实上，它的功能十分强大。能兼容许多自定义操作。  
例如：  
1. 只修改内核部分而不干扰其它（比如可以保留Magisk）  
2. 自定义内核安装（例如让用户选择安装时的操作）
3. 自动修补vbmeta（取决于你是否开启此功能）
等   
而这篇教程只聚焦于基础，不涉及更复杂的内容。所以，放心，没有什么高难度内容。 
使用AnyKernel 3的前提是您的设备具有自定义RECOVERY(例如：TWRP，OrangeFox，或者支持adb sideload)。如果不具备此条件，您可以尝试解包/打包boot.img来进行内核替换。如果您拥有SU权限，则可以使用内核刷写APP然后使用AnyKernel 3刷机。(例如：Kernel Flasher，Franco Kernel Manager，EX Kernel Manager。)       
  
***  

## 2.使用AnyKernel 3 
在电脑使用如下命令clone此项目：  
```
git clone https://github.com/osm0sis/AnyKernel3.git  
``` 
然后进入AnyKernel 3的根目录：  
接下来我们来了解AnyKernel 3的基本结构目录：
```
META-INF  
modules  
patch  
ramdisk  
tools  
LICENSE  
README.md  
anykernel.sh
```  
作为初学者，我们先不去关心其他的文件，只用关心**anykernel.sh**和**LICENSE**。    
先用你喜欢的文本编辑器打开**anykernel.sh**，可以看到很多内容，别急我先把它精简，  
精简后的内容如下：    
```
### AnyKernel3 Ramdisk Mod Script
## osm0sis @ xda-developers

### AnyKernel setup
# global properties
properties() { '
kernel.string=ExampleKernel by osm0sis @ xda-developers
do.devicecheck=1
do.modules=0
do.systemless=1
do.cleanup=1
do.cleanuponabort=0
device.name1=
device.name2=
device.name3=
device.name4=
device.name5=
supported.versions=
supported.patchlevels=
supported.vendorpatchlevels=
'; } # end properties


### AnyKernel install
## boot files attributes
boot_attributes() {
set_perm_recursive 0 0 755 644 $ramdisk/*;
set_perm_recursive 0 0 750 750 $ramdisk/init* $ramdisk/sbin;
} # end attributes

# boot shell variables
block= XXXX ;
is_slot_device=0;
ramdisk_compression=auto;
patch_vbmeta_flag=auto;

# import functions/variables and setup patching - see for reference (DO NOT REMOVE)
. tools/ak3-core.sh;

# boot install
dump_boot; # use split_boot to skip ramdisk unpack, e.g. for devices with init_boot ramdisk

write_boot; # use flash_boot to skip ramdisk repack, e.g. for devices with init_boot ramdisk
## end boot install

```  
那么这里介绍你需要了解的部分：  
1. kernel.string= ： 这里需要你填写内核的名称，你可以随意写。
2. do.devicecheck= ： 填写1开启，0是关闭。这里是用来检测设备名称的，和下面device.name1~5形成对应。如果设备名称不对应，就无法刷入进设备。  
3. device.name1 ~5 ： 这里可以填写你的设备名称。注意，有些手机是开发代号，有的是设备名称具体依靠你自行判断。使用这些参数需要先开启do.devicecheck。  
4. block= XXXX ： 需要刷写内核的分区，一般是boot。  
5. is_slot_device= ： 填写1开启，0是关闭，auto是自动判断。设备使用是使用AB插槽。一般是VAB分区或者AB分区的则需要开启。  
6. patch_vbmeta_flag= ： 填写1开启，0是关闭，auto是自动判断。是否修补vbmeta来关闭AVB验证。这项需要看设备，所以具体依靠你自行判断。    
知道了这些，你就可以用下面的通用脚本来自行修改属于你的内核刷机包了。  
通用脚本：  
```
### AnyKernel3 Ramdisk Mod Script
## osm0sis @ xda-developers

### AnyKernel setup
# global properties
properties() { '
kernel.string=ExampleKernel by osm0sis @ xda-developers # 请自行填写内核名字
do.devicecheck=0 # 关闭了设备检测
do.modules=0
do.systemless=1
do.cleanup=1
do.cleanuponabort=0
supported.versions=
supported.patchlevels=
supported.vendorpatchlevels=
'; } # end properties


### AnyKernel install
## boot files attributes
boot_attributes() {
set_perm_recursive 0 0 755 644 $ramdisk/*;
set_perm_recursive 0 0 750 750 $ramdisk/init* $ramdisk/sbin;
} # end attributes

# boot shell variables
block=boot;  # 刷写分区是boot
is_slot_device=auto; # 自动检测是否AB插槽
ramdisk_compression=auto;
patch_vbmeta_flag=auto;

# import functions/variables and setup patching - see for reference (DO NOT REMOVE)
. tools/ak3-core.sh;

# boot install
dump_boot; # use split_boot to skip ramdisk unpack, e.g. for devices with init_boot ramdisk

write_boot; # use flash_boot to skip ramdisk repack, e.g. for devices with init_boot ramdisk
## end boot install
```  
**注意这个脚本仅仅是一个可用的模板，其中内容仍然需要你按照上述内容进行修改并检查！**  

***  
  
## 3.打包你的内核  
复制你的内核进入AnyKernel 3根目录，假设你的内核名是**Image.gz**。  
那么此时你的AnyKernel 3的根目录应该看上是这样的：  
```
META-INF  
modules  
patch  
ramdisk  
tools  
LICENSE  
README.md  
anykernel.sh
Image.gz
```  
之后用zip工具压缩根目录下所有内容。请记住，**是zip**，不是rar，7z，tar等！！！    
然后你就可以用TWRP刷入并测试内核了。  
关于**LICENSE**：  
**LICENSE**是许可证，这意味着AnyKernel 3是受到法律保护的，你需要按照法律许可内行动。请严格遵守此**LICENSE**的限制。
    
***  
  
## 4.结尾
这只是一个基本使用AnyKernel 3的教程，如果你有更好的意见和建议，请在评论区不吝赐教。   

## 5.感谢  
@osm0sis
> 作者主页：[osm0sis](https://github.com/osm0sis)

**Comment by osm0sis** [comment]
Nice work, a little redundant having the example script twice, but nicely explained for beginners. 👍

You should repost in English as well. 😘

One thing:
![Screenshot_20230913-091047_Translate](https://github.com/tiann/KernelSU/assets/2858716/6fc182ab-bf64-4c30-adf2-b1444cd1ff7b)

AK3 can also be flashed while booted, using Kernel Flasher, FKM, or EXKM. So doesn't *require* custom recovery, as long as you're rooted. 🙂

**Reply by Coconutat** [reply]
Thank you for your suggestion. I will try to translate an English version tomorrow. And Thank you for your great project, which is very useful.

**Comment by xuxin3101** [comment]
我看很多教程都是说放Image.gz-dtb文件，只有这个教程说的是Image.gz文件，想问下，这有什么区别？

**Reply by Coconutat** [reply]
不同设备要的内核不一样。有的设备是Image.gz，有的是Image.gz-dtb。更详细的内容阅读官方文档：https://source.android.com/docs/core/architecture/bootloader/dtb-images?hl=zh-cn

**Comment by a289672082** [comment]
这个教程最好别看,不看你可能会用,看了就用不来了
